rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/SpatialMatcher.scala

```diff
@@ -327,6 +329,38 @@ object SpatialMatcher extends SpatialMatcherInstances {
     } yield Unit
   }
 
+  private def guard(predicate: => Boolean): OptionalFreeMapWithCost[Unit] =
+    if (predicate) OptionalFreeMapWithCost.pure(()) else OptionalFreeMapWithCost.emptyMap
+
+  private def isolateState[S, F[_]: Monad](f: StateT[F, S, _]): StateT[F, S, S] =
+    for {
+      initState   <- StateT.get[F, S]
+      _           <- f
+      resultState <- StateT.get[F, S]
+      _           <- StateT.set[F, S](initState)
+    } yield resultState
+
+  def memoizeInHashMap[A, B, C](f: (A, B) => C): (A, B) => C = {
+    val memo = mutable.HashMap[(A, B), C]()
+    (a, b) =>
+      memo.getOrElseUpdate((a, b), f(a, b))
```

Comment:
Is this piece of code executed concurrently from multiple threads?

Reply:
No, it's not.